web3/eth.py

```diff
@@ -49,6 +51,147 @@
+        logs = list()
+        while current_batch_start < current_batch_end:
+            filter_params["fromBlock"] = current_batch_start
+            filter_params["toBlock"] = current_batch_end
+            log_list = None
+            for currentTry in range(1, (retries + 1)):
+                try:
+                    logger.debug(f'Trying to get logs with params {filter_params} '
+                                 f'from block {current_batch_start} to block {current_batch_end}. '
+                                 f'(attempt {currentTry})')
+                    log_list = self.web3.manager.request_blocking(
+                        "eth_getLogs", [filter_params],
+                    )
+                    break
+                except ValueError as e:
+                    logger.warning(f'Error trying to get logs with params {filter_params} '
+                                   f'from block {current_batch_start} to block {current_batch_end}', e)
+            if log_list is None:
+                raise ValueError(f'Error trying to get logs with params {filter_params} '
+                                 f'from block {current_batch_start} to block {current_batch_end}')
+            else:
+                for log in log_list:
+                    logs.append(log)
+                current_batch_start = current_batch_end + 1
+                current_batch_end = EthGetLogsCache.get_next_batch_end(current_batch_start, to_block, batch_size)
+
+        return logs
+
+    @staticmethod
+    def generate_cache_key(address: str, topics: List[str]) -> int:
+        return hash(f"logs_{address}_{topics}")
+
+    def get(self, filter_params: Any, earliest_block_number: int, latest_block_number: int) -> List:
+        from_block: int = filter_params["fromBlock"] if "fromBlock" in filter_params else earliest_block_number
+        to_block: int = filter_params["toBlock"] if "toBlock" in filter_params else latest_block_number
+        address: str = filter_params["address"] if "address" in filter_params else ""
+        topics: List[str] = filter_params["topics"] if "topics" in filter_params else ""
+        invalidate_cache: bool = filter_params["invalidate_cache"] if "invalidate_cache" in filter_params else False
+        if "invalidate_cache" in filter_params:
+            del filter_params["invalidate_cache"]
+        # check the cache before populating with data
+        cache_key: int = self.generate_cache_key(address, topics)
+        logger.debug(f"getting logs with cache key {cache_key}")
+        if cache_key in self.logs and not invalidate_cache:
+            logger.debug(f"cache exists and is not being invalidated")
+            # here we have all the logs from earliest to latest on the last iteration
+            # we need to check if the current iteration doesn't need an update from
+            # latest of last iteration to current block
+            cached_value: EthGetLogsCachedValue = self.logs[cache_key]
```

Comment:
why check that the key is in the dictionary, and then retrieve its value? do a `get` directly, but do a short-circuit by evaluating `invalidate_cache` first